### PR TITLE
Use -Wswitch for TxoutType where possible

### PR DESCRIPTION
**Pull request by maflcko**
This removes unused `default:` cases for all `switch` statements on `TxoutType` and adds the cases (`MULTISIG`, `NULL_DATA`, `NONSTANDARD`) to `ExtractDestination` for clarity.

Also, the compiler is now able to use `-Wswitch`.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK: `-Wswitch` is good, and termination via `assert(false);` is better than UB :)

From the standard: "Flowing off the end of a function is equivalent to a return with no value; this results in undefined behavior in a value-returning function." (C++11,  [stmt.return])

**Comment by jonatack** [comment]
Concept ACK

**Comment by hebasto** [comment]
Concept ACK

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -934,9 +934,9 @@ static std::string RecurseImportData(const CScript& script, ImportData& import_d
     case TxoutType::NONSTANDARD:
     case TxoutType::WITNESS_UNKNOWN:
     case TxoutType::WITNESS_V1_TAPROOT:
-    default:
         return "unrecognized script";
-    }
+    } // no default case, so the compiler can warn about missing cases
+    CHECK_NONFATAL(false);
```
Comment (hebasto) [inline src/wallet/rpcdump.cpp:939]: I understand that RPC code shouldn't terminate the server regardless of user's input. But `TxoutType` is an `enum class`, and its set of values does not depends on user's input. Therefore, I think the usual `assert(false);` should be used here.

Comment (MarcoFalke) [reply src/wallet/rpcdump.cpp:939]: What is the difference in practise? This is dead code either way, so it doesn't matter what it does as long as it ensures that the function doesn't return normally when the dead code is executed (for whatever reason).

Comment (hebasto) [reply src/wallet/rpcdump.cpp:939]: The only difference is a uniform approach through the repo.

**Review by hebasto — APPROVED** [review]
ACK fa650ca7f19307a9237e64ac311488c8947fc12a, I have reviewed the code and it looks OK, I agree it can be merged.

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/wallet/rpcdump.cpp:939]: The rpc code has a linter to check for a uniform `CHECK_NONFATAL` in all rpc code ;)

**Review by mjdietzx — APPROVED** [review]
Ack https://github.com/bitcoin/bitcoin/pull/20211/commits/fa650ca7f19307a9237e64ac311488c8947fc12a

We do `switch (dest.which()) {` in a few places as well. This seems pretty similar, and I'm wondering if this should be refactored as well? (Probably in a different PR though?)

In https://github.com/bitcoin/bitcoin/pull/20286 I added `TxoutType& typeRet` to `ExtractDestination` [here](https://github.com/bitcoin/bitcoin/pull/20286/files#diff-4c44c096ea0d1ded9679ea69831cdd3879cb50c5cb829b20c740d96e17f53c8eR240), and I was wondering if this would allow us to make these `switch (dest.which()) {` statements more readable by doing it on `TxoutType` as done in this PR. Thoughts, and once https://github.com/bitcoin/bitcoin/pull/20286 is merged, is this something that should be done?

**Comment by MarcoFalke** [comment]
> We do switch (dest.which()) { in a few places as well. This seems pretty similar, and I'm wondering if this should be refactored as well? (Probably in a different PR though?)

What do you suggest? `which` decays the `enum class` into an integer. The compiler can't do `-Wswitch` static analysis on those switch statements.

**Comment by mjdietzx** [comment]
Well once we have (would require https://github.com/bitcoin/bitcoin/pull/20286):

```c++
bool ExtractDestination(const CScript& scriptPubKey, TxoutType& typeRet, CTxDestination& addressRet);
```

Could we do something like:
```c++
TxoutType type;
CTxDestination dest;
ExtractDestination(m_script, type, dest);
switch (type) {
  case TxoutType::PUBKEYHASH:
  case TxoutType::SCRIPTHASH: return OutputType::LEGACY;
  case TxoutType::WITNESS_V0_SCRIPTHASH:
  case TxoutType::WITNESS_V0_KEYHASH:
  case TxoutType::WITNESS_V1_TAPROOT:
  case TxoutType::WITNESS_UNKNOWN: return OutputType::BECH32;
  case TxoutType::PUBKEY:
  case TxoutType::MULTISIG:
  case TxoutType::NONSTANDARD:
  case TxoutType::NULL_DATA: return nullopt;
}
```

Note: I messed around with this, and probably made some mistakes here, but didn't see any unit tests or functional tests fail. So maybe some test coverage is needed here as well?

**Comment by MarcoFalke** [comment]
> would require #20286

Let's discuss this in the other pull then

**Comment by practicalswift** [comment]
cr ACK fa650ca7f19307a9237e64ac311488c8947fc12a: patch looks correct and `assert(false);` is better than UB :)

**Pull request merged**